### DISCUSSION TITLE
_LiteralRoundTripFixture tests failing in Firebird (need COMMIT after DDL?)

### DISCUSSION THREAD

**Discussion by fdcastel** [sqlalchemy/sqlalchemy#9593] [Usage Questions]
I'm trying to upgrade the [external Firebird Dialect](https://github.com/pauldex/sqlalchemy-firebird) to SQLAlchemy 2.0.

Running the test suite I'm getting several tests failing with a similar pattern:

```
./test/test_suite.py::BooleanTest_firebird+fdb_4_0_2816_firebird::test_render_literal_bool Failed: [undefined]sqlalchemy.exc.DatabaseError: (fdb.fbcore.DatabaseError) ('Error while preparing SQL statement:\n- SQLCODE: -204\n- Dynamic SQL Error\n- SQL error code = -204\n- Table unknown\n- T\n- At line 1, column 13', -204, 335544569)
[SQL: INSERT INTO t (x) VALUES (true)]
```

The [code for _LiteralRoundTripFixture.literal_round_trip](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/testing/suite/test_types.py#L58) creates a table and immediately try to insert some rows:

```python
            t = Table("t", metadata, Column("x", type_))
            t.create(connection)

            for value in input_:
                ins = t.insert().values(
                    x=literal(value, type_, literal_execute=True)
                )
                connection.execute(ins)
```

However, Firebird needs a `COMMIT` between DDL and DML. Without this the `INSERT` operation fails with a `Table unknown` error.

What should be the correct way to handle this in the external dialect?

**Comment by zzzeek** [comment]
we would need to know about all such tests and change them on this end

**Reply by fdcastel** [reply]
> we would need to know about all such tests and change them on this end

Sorry. I don't follow. These are tests from SQLAlchemy test suite.

I was expecting that some other dialect already had this problem and could give me some pointers about how to solve.

Currently there are NO databases which need a COMMIT between a DDL and DML?

**Reply by CaselIT** [reply]
> Currently there are NO databases which need a COMMIT between a DDL and DML?

Yes, no dialect reported such an issue. Don't know if they just choose to ignore the errors or if the db in general support this

**Reply by zzzeek** [reply]
firebird was a very difficult database to support as it has many arcane restrictions such as this one which is why we ejected it from SQLAlchemy proper.  I thought it has a maintainer though so I'm surprised to be just hearing about these issues now

**Reply by zzzeek** [reply]
one way you could work around on the dialect side is to add an after_cursor_execute() listener that just runs COMMIT after any detected DDL.   I would not make this part of the dialect though, just something that's attached during the test suite

**Reply by fdcastel** [reply]
> Yes, no dialect reported such an issue. Don't know if they just choose to ignore the errors or if the db in general support this

Damn 😞 

Every Firebird operation MUST run in the context of a Transaction. Even for READ-ONLY queries (`SELECT`s).

I'm learning both SQLAlchemy and SQLAlchemy-Firebird dialect right now. I will investigate what workarounds were used for this. But surely some SQLAlchemy support in this area would be useful.

Suggestions, of course, are welcome too.

**Reply by CaselIT** [reply]
> one way you could work around on the dialect side is to add an after_cursor_execute() listener that just runs COMMIT after any detected DDL. I would not make this part of the dialect though, just something that's attached during the test suite

that's probably the best option. Alternatively you could use Table/Metadata `after_create` event.
In any case you should use the provisions to register it.

**Reply by CaselIT** [reply]
by provision I mean these
https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/testing/provision.py
example on oracle
https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/oracle/provision.py

**Reply by zzzeek** [reply]
> > Yes, no dialect reported such an issue. Don't know if they just choose to ignore the errors or if the db in general support this
> 
> Damn disappointed
> 
> Every Firebird operation MUST run in the context of a Transaction. Even for READ-ONLY queries (`SELECT`s).

All DBAPI operations run in transactions.  what you are suggesting however is that DML is not allowed in an ongoing transaction that just had DDL within it.  that's not a typical limitation.   the workaround is to add an event hook that will emit the "COMMIT" command on the connection after any DDL statements.   the DBAPI will then autobegin for the next statement.

**Reply by fdcastel** [reply]
> firebird was a very difficult database to support as it has many arcane restrictions such as this one which is why we ejected it from SQLAlchemy proper. 

Sad to hear this. I was under the impression that it was ejected just because it wasn't mainstream (i.e. less users than Postgres/MySql/SqlServer).

It is a very good engine, really. This restriction actually is from design: it was one of the first databases to use Multiversion Concurrency Control. Even `SELECT`s need a transaction because it will determine what version of data you will see.

But, sure, I agree that it is a strange concept to anyone who came from another RDBMS.



> I thought it has a maintainer though so I'm surprised to be just hearing about these issues now

Just to be clear: the dialect WORKS. 

The problem is with the TESTs (which were made by SQLAlchemy team and, of course, not aware of this need from Firebird).

Unfortunately the current dialect is in a dire state. It's essentially a very old codebase over an old driver (fdb). The discussions on github project page are also almost inexistent 😢. 

I'm using the tests to both learn the code while trying to upgrade the project to SQLAlchemy 2.0. 

There is also a [new driver](https://pypi.org/project/firebird-driver/) much more modern which I intend to use for a new dialect as soon as I cut my teeth in SQLAlchemy (Yeah... sorry, noob here 😄).

**Reply by CaselIT** [reply]
just FYI a single dialect may support more than one driver, so you could add support to the new driver in the current dialect.
That said I don't know the maintenance status of the current one, but maybe you could give a hand as an additional maintainer

**Reply by CaselIT** [reply]
Also by all mean feel free to ask question here or on gitter https://app.gitter.im/#/room/#sqlalchemy_devel:gitter.im

**Reply by zzzeek** [reply]
yeah fdb was not the first DBAPI supported, it was doing kinterbasdb before that and I thought it still did both

**Comment by fdcastel** [comment]
Thanks for the pointers, guys. I managed to add `after_create`/`after_drop` listeners to `provision.py` which solved the `COMMIT` problem. 👍 

Now, back to [this code](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/testing/suite/test_types.py#L58):

```python
            t = Table("t", metadata, Column("x", type_))
            t.create(connection)

            for value in input_:
                ins = t.insert().values(
                    x=literal(value, type_, literal_execute=True)
                )
                connection.execute(ins)
```

I didn't found any `DROP TABLE` for this `T` table created for this test.

- Is this `t.create(connection)` supposed to execute without errors if the table already exists?
- If not, where is the code responsible for its `DROP`?

I'm asking this because sometimes I got a `Table T already exists` in my tests.

**Reply by fdcastel** [reply]
My fault. I just found `drop_all_tables_from_metadata` from `util.py` (called from `fixtures.py`). 😅

My real problem was caused by wrong column types (old dialect creating `SMALLINT`s instead of `BOOLEAN`s, natively supported by Firebird since v3.0)

All fixed now! 🙌

**Reply by CaselIT** [reply]
I think in some cases the table creation is bot committed, so there is no drop required. Since transactional ddl is not supported by all db there is also the drop all call
